Github Action improvements

erikd opened the pull request:
# Changelog

```yaml
- description: Github Action improvements
  type:
    - maintenance    # not directly related to the code
```

# Context

Improve and simplify GHA.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

<!-- 
### Note on CI ###
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges. Please contact IOG node developers to do this
for you. 
-->

erikd committed:
CI: Simplify Github Actions

  * Make it more like `cardano-base`.
  * Drop the ad-hoc caching in favour of Cabal caching.

carbolymer commented on .github/workflows/haskell.yml:
```diff
@@ -119,26 +99,7 @@ jobs:
         KEEP_WORKSPACE: 1
-      run: cabal test all
-
-    - # Delete golden files and regenerate them in the next step to ensure
```
I think it may be useful to have that. That's a safeguard against putting non-golden files into golden directory accidentaly.

erikd replied:
Restored.

erikd committed:
CI: Actually use the cache dependencies

Previously the GHAs would store the cache after building the
dependencies but never restore them. Did a thorough investigation
of the issue and added proper comments about how it works and why.

carbolymer commented on .github/workflows/haskell.yml:
```diff
@@ -66,60 +66,53 @@ jobs:
+          ${{ steps.setup-haskell.outputs.cabal-store }}
+          dist-newstyle
+        key: cache-${{ env.CABAL_CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+        restore-keys: |
```
What's the purpose of of `restore-keys` here? Shouldn't `key` be enough?

erikd replied:
Not sure. THis cache restoration code was cargo culted from elsewhere. I can try deleting it if you like, see if it still works.

erikd replied:
Left over garbage from where I cargo-culted this from.

PR to remove it at:  https://github.com/IntersectMBO/cardano-cli/pull/628
